Some tasks solved

DDRDmakar opened the pull request.

DDRDmakar committed:
Update Simple.kt

DDRDmakar committed:
Update IfElse.kt

ice-phoenix commented:
recheck assignment

mglukhikh commented:
recheck assignment

mglukhikh commented:
У вас файл Simple.kt содержит какой-то кусок кода на Java, который, естественно, не компилируется и проверить его из-за этого нельзя. Прошу исправить проблему

DDRDmakar commented:
Спасибо, попробую исправить.

DDRDmakar committed:
Bugs in simple file fixed

mglukhikh commented:
Обратите внимание, что задача ageDescription решена с ошибками, некоторые тесты не проходят. Остальное, насколько я понимаю, правильно.

DDRDmakar committed:
Lesson 2 task 1

DDRDmakar committed:
Lesson 2 task 1 fixes

maximpetrov commented:
recheck assignment

maximpetrov commented:
recheck assignment

DDRDmakar committed:
lesson 1 task 1 - fixed bug with data types

gagarski commented:
Тесты для `angleInRadian` проваливаются из-за недостаточной точности числа π. Вы можете использовать константу из класса `java.lang.Math` (см. часть 1 учебного пособия).

mglukhikh commented:
Задача ageDescription не всегда выдаёт верный ответ

mglukhikh commented:
Остальные (кроме angleInRadian и ageDescription) решения верные

Commit:
Some loop functions done ( Bugs are not fixed yet )
	modified:   src/lesson1/task1/Simple.kt
	modified:   src/lesson2/task2/Logical.kt
	modified:   src/lesson3/task1/Loop.kt

DDRDmakar committed:
Loop ( Some bugs fixed, but uncompilable yet )

Commit:
And again a little bit more bugs fixed ( uncompilable version )
	modified:   src/lesson3/task1/Loop.kt

DDRDmakar committed:
Loop ( All bugs fixed )

ice-phoenix commented on src/lesson1/task1/Simple.kt:
```diff
@@ -1,57 +1,57 @@
 @file:Suppress("UNUSED_PARAMETER")
-package lesson1.task1
+        package lesson1.task1
```
Отформатируйте код при помощи встроенной функции форматирования IntelliJ IDEA и в будущем постарайтесь придерживаться такого стиля кодирования.

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -1,39 +1,47 @@
 @file:Suppress("UNUSED_PARAMETER")
-package lesson2.task1
+        package lesson2.task1
```
Форматирование кода.

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -157,7 +299,17 @@ fun hasDifferentDigits(n: Int): Boolean = TODO()
  * Например, 2-я цифра равна 4, 7-я 5, 12-я 6.
  */
-fun squareSequenceDigit(n: Int): Int = TODO()
+fun squareSequenceDigit(n: Int): Int
```
Оцените, строку какого размера придется создать вашему коду в случае, если `n == 10^6`. Какой объем памяти потребуется для ее хранения, если считать, что хранение одного символа требует 1 байта?

Можно ли изменить ваш код так, чтобы не требовалось сохранять все элементы последовательности квадратов чисел в одну строку?

ice-phoenix commented:
recheck all

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,30 +57,85 @@ fun digitCountInNumber(n: Int, m: Int): Int =
+	if ( m > n ) i = m
+	else i = n
+	
+	while ( i % m != 0 || i % n != 0 ) i ++
```
Для решения этой задачи рекомендуется воспользоваться известными алгоритмами нахождения LCM (см. интернет)

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -140,7 +257,32 @@ fun revert(n: Int): Int = TODO()
+	// while ( Math.pow ( 10, n_digits.toDouble() ) <= n ) ++ n_digits
+	//n_digits = Math.log ( n.toDouble() ).toInt()
+
+	var current_number_line : String = "$n"
```
- Данная переменная может быть объявлена как `val`
- Тип переменной можно не указывать, система типов Котлина сама выведет нужный тип

Commit:
Last task in loop fixed
	modified:   src/lesson3/task1/Loop.kt

DDRDmakar committed:
lesson 3 task 1 changes ( compilable version )

ice-phoenix commented:
Просьба исправить все указанные выше замечания к форматированию кода.

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -1,39 +1,47 @@
+       // String answer
+       // Int last_digit
+        val last_digit = age % 10
+        if ( last_digit == 0 || last_digit > 4 || ( age >= 10 && age <= 20 ) ) return "$age лет"
```
Данное выражение лучше оформить как `when`

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -56,7 +80,23 @@ fun timeForHalfWay(t1: Double, v1: Double,
+        var R : Int
+
+        if
+        (       ( kingX == rookX1 || kingY == rookY1 ) &&
```
Выражения, соответствующие фактам атаки короля той или иной фигурой, следует вынести в отдельные переменные.

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -89,4 +129,23 @@ fun triangleKind(a: Double, b: Double, c: Double): Int = TODO()
+        var R : Int
+
+        if ( c > b || a > d ) R = -1
+        else if ( a <= c && c <= b && b <= d ) R = b - c
```
Данное выражение лучше оформить как `when`

ice-phoenix commented on src/lesson2/task2/Logical.kt:
```diff
@@ -17,15 +17,34 @@ fun pointInsideCircle(x: Double, y: Double, x0: Double, y0: Double, r: Double) =
+    s1 = ( ( number - number % 1000 ) / 1000 ) + ( ( number % 1000 - number % 100 ) / 100 )
+    s2 = ( ( number % 100 - number % 10 ) / 10 ) + ( number % 10 )
+    
+    if ( s1 == s2 ) return true
```
Данное выражение лучше оформить как `return s1 == s2`

ice-phoenix commented on src/lesson2/task2/Logical.kt:
```diff
@@ -17,15 +17,34 @@ fun pointInsideCircle(x: Double, y: Double, x0: Double, y0: Double, r: Double) =
+fun queenThreatens(x1: Int, y1: Int, x2: Int, y2: Int): Boolean
+{
+    if
+    (
```
Данное выражение лучше оформить как `return [if condition]`

ice-phoenix commented on src/lesson2/task2/Logical.kt:
```diff
@@ -34,8 +53,11 @@ fun queenThreatens(x1: Int, y1: Int, x2: Int, y2: Int): Boolean = TODO()
-                 x2: Double, y2: Double, r2: Double): Boolean = TODO()
+fun circleInside(x1: Double, y1: Double, r1: Double, x2: Double, y2: Double, r2: Double): Boolean
+{   // Если расстояние между центрами + радиус первого меньше радиуса второго
+    if ( Math.sqrt ( ( x2 - x1 ) * ( x2 - x1 ) + ( y2 - y1 ) * ( y2 - y1 ) ) + r1 <= r2 ) return true
```
Данное выражение лучше оформить как `return [if condition]`

ice-phoenix commented on src/lesson2/task2/Logical.kt:
```diff
@@ -46,4 +68,13 @@ fun circleInside(x1: Double, y1: Double, r1: Double,
+{
+    if
+    (
+		( a <= r && b <= s ) || ( b <= r && a <= s ) ||
```
Данное выражение лучше оформить как `return [if condition]`

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -20,11 +21,12 @@ fun factorial(n: Int): Double {
-    return true
+	if (n < 2) return false
+	if (n == 2) return true
+	for (m in 2..n-1) {
```
Нет ли более точной оценки сверху для максимального возможного делителя `n`?

ice-phoenix replied:
Например, как в вашем предыдущем решении

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -48,39 +50,87 @@ fun isPerfect(n: Int): Boolean {
+	if (m > n) i = m
+	else i = n
+
+	while (i % m != 0 || i % n != 0) i++
```
Нет ли более быстрых алгоритмов поиска LCM?

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -48,39 +50,87 @@ fun isPerfect(n: Int): Boolean {
  */
-fun minDivisor(n: Int): Int = TODO()
+fun minDivisor(n: Int): Int {
+	for (i in 2..n) {
```
Каким числом ограничен максимально возможный делитель числа `n`, не равный `n`?

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -140,7 +246,30 @@ fun revert(n: Int): Int = TODO()
+
+	var current_number_line: String = "$n"
+
+	for (i in 0..current_number_line.length - 1) {
```
Можно ли данную задачу решить с помощью других функций данного раздела?

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -156,8 +285,22 @@ fun hasDifferentDigits(n: Int): Boolean = TODO()
  */
-fun squareSequenceDigit(n: Int): Int = TODO()
+fun squareSequenceDigit(n: Int): Int {
+	var stopka: String = ""
```
Не следует называть переменные транслитом

DDRDmakar committed:
Code formatting + some optimization done ( compilable version )

DDRDmakar commented:
Сделал изменения, как вы просили, но бот вылетает по времени. Похоже, что при проверке isPrime() Эх

gagarski commented:
@DDRDmakar про `isPrime()` вы догадались верно, но непонятно, зачем вы вообще его изначально правили (это был пример). Теперь он работает дольше (да ещё и с ошибкой).

DDRDmakar committed:
Loop ( isPrime() repaired )

DDRDmakar committed:
Part of lesson 4 done ( compilable version )

DDRDmakar commented:
Похоже, при проверке функции digitNumber() происходит переход чисел из signed в unsigned.
Функция digitNumber(-2147483647) находит 10 разрядов, как и требуется
А digitNumber(-2147483648) находит один разряд ( то есть, это число преобразуется в ноль ).

DDRDmakar commented:
А, всё, понял. Простите. Я его в положительное переводил.

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -128,7 +139,15 @@ fun center(list: MutableList<Double>): MutableList<Double> = TODO()
-fun times(a: List<Double>, b: List<Double>): Double = TODO()
+fun times(a: List<Double>, b: List<Double>): Double {
+    var qbzv : Double = 0.0
+    if ( ( a.size != b.size ) || a.isEmpty() ) return 0.0
```
Проверка `a.isEmpty()` здесь излишняя

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -104,22 +104,33 @@ fun buildSumExample(list: List<Int>) = list.joinToString(separator = " + ", post
-fun mean(list: List<Double>): Double = TODO()
+fun mean(list: List<Double>): Double {
+    if ( list.isEmpty() ) return 0.0
+    else return ( list.sum().toDouble() / list.size.toDouble() )
```
Вызывать `toDouble()` здесь необязательно

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -157,7 +194,32 @@ fun accumulate(list: MutableList<Double>): MutableList<Double> = TODO()
-fun factorize(n: Int): List<Int> = TODO()
+fun number_is_prime(n: Int): Boolean {
+    if ( n == 0 || n == 1 ) return false
+    for( i in 2..Math.abs( n/2 )) if ( n % i == 0 ) return false
```
Зачем `Math.abs(n / 2)`?

DDRDmakar replied:
Чтобы цикл запускался, когда аргумент отрицательный.

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -157,7 +194,32 @@ fun accumulate(list: MutableList<Double>): MutableList<Double> = TODO()
+
+fun factorize(n: Int): List<Int> {
+    var current_number: Int = n
+    var prime_divisors_storage: MutableList <Int> = mutableListOf()
```
Лучше `var prime_divisors_storage = mutableListOf<Int>()`

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -157,7 +194,32 @@ fun accumulate(list: MutableList<Double>): MutableList<Double> = TODO()
+
+    do {
+        for (i in current_number / 2 downTo 2) {
+            if (number_is_prime(i) && current_number % i == 0) {
```
Подумайте, как можно реализовать перебор делимости числа на различные `простые` делители более эффективным образом

DDRDmakar committed:
List done ( except of "russian" ) - Compilable version

DDRDmakar committed:
Bugs in '''List'' fixed - Compilable version

ice-phoenix commented:
recheck all

ice-phoenix commented on src/lesson1/task1/Simple.kt:
```diff
@@ -102,12 +103,12 @@ fun travelMinutes(hoursDepart: Int, minutesDepart: Int, hoursArrive: Int, minute
  * Например, 100 рублей под 10% годовых превратятся в 133.1 рубля
  */
-fun accountInThreeYears(initial: Int, percent: Int): Double = TODO()
+fun accountInThreeYears(initial: Int, percent: Int): Double = initial * (1.0 + (percent / 100.0)) * (1.0 + (percent / 100.0)) * (1.0 + (percent / 100.0)) * 1.0
```
Повторяющиеся подвыражения лучше вынести в отдельные переменные

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -33,7 +34,18 @@ fun minBiRoot(a: Double, b: Double, c: Double): Double {
+
+    // String answer
+    // Int last_digit
+    val last_digit = age % 10
```
Почему в отдельную переменную вынесли только `age % 10`, а `age % 100` оставили как есть?

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -56,7 +82,18 @@ fun timeForHalfWay(t1: Double, v1: Double,
-                       rookX2: Int, rookY2: Int): Int = TODO()
+                       rookX2: Int, rookY2: Int): Int {
+
+    val r1: Boolean = (kingX == rookX1 || kingY == rookY1)
```
- Система типов Котлина выведет тип переменной автоматически

ice-phoenix commented on src/lesson2/task2/Logical.kt:
```diff
@@ -17,15 +18,26 @@ fun pointInsideCircle(x: Double, y: Double, x0: Double, y0: Double, r: Double) =
  */
-fun isNumberHappy(number: Int): Boolean = TODO()
+fun isNumberHappy(number: Int): Boolean {
+    val s1: Int = ((number - number % 1000) / 1000) + ((number % 1000 - number % 100) / 100)
```
- Система типов Котлина выведет тип переменной автоматически

ice-phoenix commented on src/lesson2/task2/Logical.kt:
```diff
@@ -46,4 +59,10 @@ fun circleInside(x1: Double, y1: Double, r1: Double,
-fun brickPasses(a: Int, b: Int, c: Int, r: Int, s: Int): Boolean = TODO()
+fun brickPasses(a: Int, b: Int, c: Int, r: Int, s: Int): Boolean {
+    return (
+        (a <= r && b <= s) || (b <= r && a <= s) ||
```
Лучше найти две минимальные стороны кирпича и все считать относительно них

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,30 +58,78 @@ fun digitCountInNumber(n: Int, m: Int): Int =
+
+    if (m > n) i = m
+    else i = n
+    val i_step: Int = i
```
Все это можно записать как `val i_step = if (m > n) m else n`

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,30 +58,78 @@ fun digitCountInNumber(n: Int, m: Int): Int =
+    val i_step: Int = i
+
+//    while (i % m != 0 || i % n != 0) i++
+    while (i % m != 0 || i % n != 0) i += i_step
```
Можно ли найти НОК более эффективным способом?

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -114,7 +163,28 @@ fun squareBetweenExists(m: Int, n: Int): Boolean = TODO()
+    var total: Double
+
+    i = 3
+    total = x
```
Можно ли, используя какие-либо тригонометрические соотношения, ограничить величину `x`, которую мы используем в разложении?

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -123,15 +193,50 @@ fun sin(x: Double, eps: Double): Double = TODO()
+    var total: Double
+
+    i = 2
+    total = 1.0
```
Можно ли, используя какие-либо тригонометрические соотношения, ограничить величину `x`, которую мы используем в разложении?

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -123,15 +193,50 @@ fun sin(x: Double, eps: Double): Double = TODO()
  */
-fun revert(n: Int): Int = TODO()
+fun revert(n: Int): Int {
+    var reverted: Int
```
[Здесь и ранее] Лучше совмещать объявление и определение переменной

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -140,7 +245,16 @@ fun revert(n: Int): Int = TODO()
+fun isPalindrome(n: Int): Boolean {
+
+    // var current_number_line: String = "$n"
+    val current_number_line: String = Integer.toString ( n )
```
- Система типов Котлина выведет тип переменной автоматически

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -140,7 +245,16 @@ fun revert(n: Int): Int = TODO()
+    // var current_number_line: String = "$n"
+    val current_number_line: String = Integer.toString ( n )
+
+    for (i in 0..current_number_line.length - 1) {
```
Можно ли проверять меньше элементов в строке для проверки на палиндром?

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -104,22 +104,33 @@ fun buildSumExample(list: List<Int>) = list.joinToString(separator = " + ", post
  */
-fun center(list: MutableList<Double>): MutableList<Double> = TODO()
+fun center(list: MutableList<Double>): MutableList<Double> {
+    val ariphmetical_average : Double = list.sum() / list.size
```
Что будет, если список пуст?

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -157,15 +193,40 @@ fun accumulate(list: MutableList<Double>): MutableList<Double> = TODO()
  * Результат разложения вернуть в виде строки, например 75 -> 3*5*5
  */
-fun factorizeToString(n: Int): String = TODO()
+fun factorizeToString(n: Int): String = factorize ( n ).toMutableList().joinToString ( "*", "", "", -1, "..." )
```
Зачем `toMutableList()`?

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -174,7 +235,21 @@ fun factorizeToString(n: Int): String = TODO()
  */
-fun convert(n: Int, base: Int): List<Int> = TODO()
+fun convert(n: Int, base: Int): List<Int> {
+    var digit_storage = mutableListOf <Int> ()
```
- Данная переменная может быть объявлена как `val`

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -174,7 +235,21 @@ fun factorizeToString(n: Int): String = TODO()
+    }
+    digit_storage = digit_storage.reversed().toMutableList()
+
+    return digit_storage
```
Лучше `digit_storage.reversed()`

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -184,7 +259,16 @@ fun convert(n: Int, base: Int): List<Int> = TODO()
  */
-fun convertToString(n: Int, base: Int): String = TODO()
+fun convertToString(n: Int, base: Int): String {
+    val digit_storage = convert ( n, base )
```
Можно ли эту функцию реализовать с использование `convert`?

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -193,7 +277,13 @@ fun convertToString(n: Int, base: Int): String = TODO()
  */
-fun decimal(digits: List<Int>, base: Int): Int = TODO()
+fun decimal(digits: List<Int>, base: Int): Int {
+    var target : Int = 0
```
- Система типов Котлина выведет тип переменной автоматически

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -204,7 +294,16 @@ fun decimal(digits: List<Int>, base: Int): Int = TODO()
  */
-fun decimalFromString(str: String, base: Int): Int = TODO()
+fun decimalFromString(str: String, base: Int): Int {
+    var target : Int = 0
```
Можно ли эту функцию реализовать при помощи функции `decimal`?

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -214,7 +313,23 @@ fun decimalFromString(str: String, base: Int): Int = TODO()
+fun roman(n: Int): String {
+    var target : String = ""
+
+    val digit_1 : List <String>    = listOf ( "", "I", "II", "III", "IV", "V", "VI", "VII", "VIII", "IX" ) // 1 2 3 4 5 6 7 8 9
```
- Система типов Котлина выведет тип переменной автоматически

Commit:
Code optimization ( lessons 1 - 4 )
	modified:   src/lesson2/task1/IfElse.kt
	modified:   src/lesson2/task2/Logical.kt
	modified:   src/lesson3/task1/Loop.kt
	modified:   src/lesson4/task1/List.kt

Commit:
Compilation errors fixed ( maybe not all )
	modified:   src/lesson1/task1/Simple.kt
	modified:   src/lesson2/task1/IfElse.kt
	modified:   src/lesson2/task2/Logical.kt
	modified:   src/lesson3/task1/Loop.kt
	modified:   src/lesson4/task1/List.kt

ice-phoenix commented on src/lesson1/task1/Simple.kt:
```diff
@@ -102,12 +103,12 @@ fun travelMinutes(hoursDepart: Int, minutesDepart: Int, hoursArrive: Int, minute
  * Например, 100 рублей под 10% годовых превратятся в 133.1 рубля
  */
-fun accountInThreeYears(initial: Int, percent: Int): Double = TODO()
+fun accountInThreeYears(initial: Int, percent: Int): Double = initial.toDouble() * Math.pow ((1.0 + (percent / 100.0)), 3.0 )
```
Можно обойтись без вызова `toDouble` в выражении `initial.toDouble()`

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,30 +58,66 @@ fun digitCountInNumber(n: Int, m: Int): Int =
+fun lcm(m: Int, n: Int): Int {
+    val s_step = if ( m > n ) m else n
+
+    for ( i in s_step .. m*n step s_step ) if ( i%m == 0 && i%n == 0 ) return i
```
Можно ли найти НОК без перебора?

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -104,22 +104,34 @@ fun buildSumExample(list: List<Int>) = list.joinToString(separator = " + ", post
  */
-fun center(list: MutableList<Double>): MutableList<Double> = TODO()
+fun center(list: MutableList<Double>): MutableList<Double> {
+	if ( list.isEmpty() ) return list
```
Форматирование кода

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -104,22 +104,34 @@ fun buildSumExample(list: List<Int>) = list.joinToString(separator = " + ", post
-fun center(list: MutableList<Double>): MutableList<Double> = TODO()
+fun center(list: MutableList<Double>): MutableList<Double> {
+	if ( list.isEmpty() ) return list
+    val ariphmetical_average = list.sum() / list.size.toDouble()
```
Можно обойтись без вызова `toDouble()`

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -114,7 +151,25 @@ fun squareBetweenExists(m: Int, n: Int): Boolean = TODO()
+    var part: Double
+
+    while (true) {
+        part = Math.pow(x, i.toDouble()) / factorial(i)
```
Подумайте, не будет ли для данной реализации проблемой работа с большими входными значениями

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -204,7 +293,14 @@ fun decimal(digits: List<Int>, base: Int): Int = TODO()
+	var digit_storage = mutableListOf <Int> ()
+	var target = 0
+	
+	for ( element in str ) digit_storage.add ( element.toInt() )
```
Посмотрите, что делает метод `Char.toInt()`

DDRDmakar committed:
Errors corrected ( 2 ).  + code formatting ( Compilable version)

Commit:
Some mistakes corrected. Code formating is not done yet
	modified:   src/lesson1/task1/Simple.kt
	modified:   src/lesson3/task1/Loop.kt
	modified:   src/lesson4/task1/List.kt

ice-phoenix commented:
recheck all

ice-phoenix commented on src/lesson2/task2/Logical.kt:
```diff
@@ -34,8 +46,9 @@ fun queenThreatens(x1: Int, y1: Int, x2: Int, y2: Int): Boolean = TODO()
-fun circleInside(x1: Double, y1: Double, r1: Double,
-                 x2: Double, y2: Double, r2: Double): Boolean = TODO()
+fun circleInside(x1: Double, y1: Double, r1: Double, x2: Double, y2: Double, r2: Double): Boolean {   // Если расстояние между центрами + радиус первого меньше радиуса второго
+    return (Math.sqrt((x2 - x1) * (x2 - x1) + (y2 - y1) * (y2 - y1)) + r1 <= r2)
```
Функцию с телом в виде одного `return expr` можно записать как `fun foo(...) = expr`

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -157,15 +195,40 @@ fun accumulate(list: MutableList<Double>): MutableList<Double> = TODO()
  * Результат разложения вернуть в виде строки, например 75 -> 3*5*5
  */
-fun factorizeToString(n: Int): String = TODO()
+fun factorizeToString(n: Int): String = factorize(n).joinToString("*", "", "", -1, "...")
```
Параметры функции, которые вы не меняете по сравнению с параметрами по умолчанию, можно не указывать

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -204,7 +294,13 @@ fun decimal(digits: List<Int>, base: Int): Int = TODO()
+fun decimalFromString(str: String, base: Int): Int {
+    val digit_storage = mutableListOf <Int>()
+
+    for (element in str) digit_storage.add((element - if ('0' <= element && element <= '9') '0' else 'a' - 10).toInt())
```
Операцию приведения символа к его цифровому представлению лучше вынести в отдельную локальную функцию

DDRDmakar committed:
Lesson 5 - 4 tasks

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -42,9 +55,23 @@ fun ageDescription(age: Int): String = TODO()
+
+    // V T
+
+    if (half <= s1) return half / v1
```
Данное выражение лучше оформить как `when`

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,30 +58,66 @@ fun digitCountInNumber(n: Int, m: Int): Int =
+fun lcm(m: Int, n: Int): Int {
+    val s_step = if (m > n) m else n
+
+    for (i in s_step..m * n step s_step) if (i % m == 0 && i % n == 0) return i
```
Нет ли более быстрых алгоритмов поиска НОК?

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -157,15 +195,40 @@ fun accumulate(list: MutableList<Double>): MutableList<Double> = TODO()
+    val prime_divisors_storage = mutableListOf <Int>()
+
+    while (true) {
+        for (i in current_number / 2 downTo 2) {
```
Как себя будет вести ваш код в случае, если `n == 1`?

ice-phoenix replied:
Почему бы не перебирать делители от минимального к максимальному, убрав необходимость проверки на простоту?

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -184,7 +259,16 @@ fun convert(n: Int, base: Int): List<Int> = TODO()
+
+    if (n < 0 || base < 2 || base > 36) return ""
+
+    for (element in digit_storage) digit_line += if (element < 10) "$element" else "${('a' + element - 10).toChar()}"
```
Для сборки строки по частям лучше использовать `StringBuilder`

ice-phoenix commented on src/lesson5/task1/Parse.kt:
```diff
@@ -60,7 +60,39 @@ fun main(args: Array<String>) {
+    }
+    catch(e: NumberFormatException) { return "" }
+
+    numeric_date += if(date_part_storage[0].length < 2 ) '0' + date_part_storage[0] else date_part_storage[0]
```
`StringBuilder`

ice-phoenix commented on src/lesson5/task1/Parse.kt:
```diff
@@ -83,7 +115,20 @@ fun dateDigitToStr(digital: String): String = TODO()
+    var index = 0
+
+    while(index < target.length - 1) {
+        if(('0' > target[index] || target[index] > '9') && target[index] != ' ' && target[index] != '-' && target[index] != '+' && target[index] != '(' && target[index] != ')') return ""
```
- Лучше завести множество (`Set`) допустимых элементов строки и проверять очередной символ на вхождение / невхождение в это множество, чем делать это через кучу проверок `char != smth`
- `return ""` должен быть на новой строке

ice-phoenix commented on src/lesson5/task1/Parse.kt:
```diff
@@ -83,7 +115,20 @@ fun dateDigitToStr(digital: String): String = TODO()
+    while(index < target.length - 1) {
+        if(('0' > target[index] || target[index] > '9') && target[index] != ' ' && target[index] != '-' && target[index] != '+' && target[index] != '(' && target[index] != ')') return ""
+        else if(('0' > target[index] || target[index] > '9') && target[index] != '+') {
+            target = target.replace( target[index].toString(), "")
```
Лучше все делать в два отдельных этапа. Сперва проверить строку на валидность, а потом (если она валидная) убрать из нее все ненужные символы разом.

ice-phoenix commented on src/lesson5/task1/Parse.kt:
```diff
@@ -95,8 +140,18 @@ fun flattenPhoneNumber(phone: String): String = TODO()
+    try {
+        for (element in jump_storage) if (element != "-" && element != "%") lucky_jump_storage.add(element.toInt())
+    }
+    catch(e: NumberFormatException) {return -1}
```
Форматирование кода

ice-phoenix commented on src/lesson5/task1/Parse.kt:
```diff
@@ -107,7 +162,32 @@ fun bestLongJump(jumps: String): Int = TODO()
-fun bestHighJump(jumps: String): Int = TODO()
+fun bestHighJump(jumps: String): Int? {
+    // Проверка на кривые символы
+    if(jumps.any{ ('0' > it || it > '9') && it != ' ' && it != '+' && it != '-' && it != '%'  } ) return -1
```
См. замечание к `flattenPhoneNumber`

ice-phoenix commented on src/lesson5/task1/Parse.kt:
```diff
@@ -107,7 +162,32 @@ fun bestLongJump(jumps: String): Int = TODO()
+    // Формирование пар
+    for( i in 0 .. jump_storage.size - 1 step 2 ) {
+        if(jump_storage[i+1].any{ '0' <= it && it <= '9' }) return -1
+        attempt_storage.add( jump_storage[i] + ' ' + jump_storage[i+1] )
```
Лучше завести `Map` из высоты в описание попыток. Это значительно упростит и дальнейший код.

DDRDmakar committed:
Lesson 5 - improved algorithms

ice-phoenix commented on src/lesson1/task1/Simple.kt:
```diff
@@ -102,12 +103,12 @@ fun travelMinutes(hoursDepart: Int, minutesDepart: Int, hoursArrive: Int, minute
  *Необходимо вывести число, полученное из заданного перестановкой цифр в обратном порядке (например, 874).
  */
-fun numberRevert(number: Int): Int = TODO()
+fun numberRevert(number: Int): Int = ((number % 10) * 100) + (((number % 100) - (number % 10))) + ((number - number % 100) / 100)
```
Лишние скобки лучше убрать для облегчения понимания выражения

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -33,7 +34,19 @@ fun minBiRoot(a: Double, b: Double, c: Double): Double {
+
+    // String answer
+    // Int last_digit
+    val last_digit = age % 10
```
Переменные в Котлине принято называть `camelCase`

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -33,7 +34,19 @@ fun minBiRoot(a: Double, b: Double, c: Double): Double {
-fun ageDescription(age: Int): String = TODO()
+fun ageDescription(age: Int): String {
+
+    // String answer
```
Что означает этот комментарий?

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -33,7 +34,19 @@ fun minBiRoot(a: Double, b: Double, c: Double): Double {
+    val last_two = age % 100
+
+    return when {
+        (last_digit == 0 || last_digit > 4 || (last_two >= 10 && last_two <= 20)) -> "$age лет"
```
Можно проще как `last_two in 10..20`

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -42,9 +55,24 @@ fun ageDescription(age: Int): String = TODO()
-                   t3: Double, v3: Double): Double = TODO()
+fun timeForHalfWay(t1: Double, v1: Double, t2: Double, v2: Double, t3: Double, v3: Double): Double {
+
+    // Double answer; s1; s2; s3; half
```
Устаревшие комментарии лучше удалять

ice-phoenix commented on src/lesson2/task2/Logical.kt:
```diff
@@ -17,15 +18,24 @@ fun pointInsideCircle(x: Double, y: Double, x0: Double, y0: Double, r: Double) =
  * Определить, угрожают ли они друг другу. Вернуть true, если угрожают.
  */
-fun queenThreatens(x1: Int, y1: Int, x2: Int, y2: Int): Boolean = TODO()
+fun queenThreatens(x1: Int, y1: Int, x2: Int, y2: Int): Boolean = (
```
Лишние скобки

ice-phoenix commented on src/lesson2/task2/Logical.kt:
```diff
@@ -34,8 +44,7 @@ fun queenThreatens(x1: Int, y1: Int, x2: Int, y2: Int): Boolean = TODO()
  */
-fun circleInside(x1: Double, y1: Double, r1: Double,
-                 x2: Double, y2: Double, r2: Double): Boolean = TODO()
+fun circleInside(x1: Double, y1: Double, r1: Double, x2: Double, y2: Double, r2: Double): Boolean = (Math.sqrt((x2 - x1) * (x2 - x1) + (y2 - y1) * (y2 - y1)) + r1 <= r2)
```
Длинные строчки принято разбивать на несколько более коротких

ice-phoenix commented on src/lesson2/task2/Logical.kt:
```diff
@@ -46,4 +55,31 @@ fun circleInside(x1: Double, y1: Double, r1: Double,
+	
+	return (min_1 <= r && min_2 <= s) || (min_1 <= s && min_2 <= r)
+	
+// 	return (
```
Устаревший комментарий

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,30 +58,63 @@ fun digitCountInNumber(n: Int, m: Int): Int =
  */
-fun digitNumber(n: Int): Int = TODO()
+fun digitNumber(n: Int): Int {
+    var digits_amount = 0
```
`camelCase`

ice-phoenix replied:
Ап

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,30 +58,63 @@ fun digitCountInNumber(n: Int, m: Int): Int =
-fun fib(n: Int): Int = TODO()
+fun fib(n: Int): Int {
+    var f = 1
+    var f_prev = 1
```
`camelCase`

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,30 +58,63 @@ fun digitCountInNumber(n: Int, m: Int): Int =
  * минимальное число k, которое делится и на m и на n без остатка
  */
-fun lcm(m: Int, n: Int): Int = TODO()
+fun gcb(m: Int, n: Int): Int = if(n == 0) m else gcb(n, m % n)
```
`gcd`

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,30 +58,63 @@ fun digitCountInNumber(n: Int, m: Int): Int =
-fun lcm(m: Int, n: Int): Int = TODO()
+fun gcb(m: Int, n: Int): Int = if(n == 0) m else gcb(n, m % n)
+
+fun lcm(m: Int, n: Int): Int = m / gcb(m, n) *  n
```
Лучше записать как `m * n / gcd(m, n)`

DDRDmakar replied:
Этот пример я нашёл в интернете, тоже думал поменять действия местами. Но если сначала умножать, а потом делить, то быстрее произойдёт переполнение переменной.

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -123,15 +175,45 @@ fun sin(x: Double, eps: Double): Double = TODO()
  */
-fun revert(n: Int): Int = TODO()
+fun revert(n: Int): Int {
+    var parameter_n = n
```
`camelCase`

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -140,7 +222,14 @@ fun revert(n: Int): Int = TODO()
  */
-fun isPalindrome(n: Int): Boolean = TODO()
+fun isPalindrome(n: Int): Boolean {
+    val current_number_line = Integer.toString(n)
```
Можно просто `n.toString()`

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -140,7 +222,14 @@ fun revert(n: Int): Int = TODO()
+    val current_number_line = Integer.toString(n)
+
+    for (i in 0..current_number_line.length / 2) {
+        if (current_number_line [i] != current_number_line [current_number_line.length - 1 - i]) return false
```
Форматирование кода

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -156,8 +245,22 @@ fun hasDifferentDigits(n: Int): Boolean = TODO()
+    var last_sequence_element = ""
+    var sequence_len = 0
+
+//	if (n == 0) return 0
```
Устаревший комментарий

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -156,8 +245,22 @@ fun hasDifferentDigits(n: Int): Boolean = TODO()
  */
-fun squareSequenceDigit(n: Int): Int = TODO()
+fun squareSequenceDigit(n: Int): Int {
+    var last_sequence_element = ""
```
`camelCase`

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -104,22 +105,34 @@ fun buildSumExample(list: List<Int>) = list.joinToString(separator = " + ", post
  */
-fun abs(v: List<Double>): Double = TODO()
+fun abs(v: List<Double>): Double {
+    // val v1 : List <Double> = v.map {it * it}
```
?

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -104,22 +105,34 @@ fun buildSumExample(list: List<Int>) = list.joinToString(separator = " + ", post
-fun abs(v: List<Double>): Double = TODO()
+fun abs(v: List<Double>): Double {
+    // val v1 : List <Double> = v.map {it * it}
+    return Math.sqrt((v.map { it * it }).sum())
```
Функцию с телом в виде одного `return expr` можно записать как `fun foo(...) = expr`

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -104,22 +105,34 @@ fun buildSumExample(list: List<Int>) = list.joinToString(separator = " + ", post
-fun center(list: MutableList<Double>): MutableList<Double> = TODO()
+fun center(list: MutableList<Double>): MutableList<Double> {
+    if (list.isEmpty()) return list
+    val ariphmetical_average = list.sum() / list.size
```
Можно было воспользоваться функцией `mean`

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -184,7 +259,16 @@ fun convert(n: Int, base: Int): List<Int> = TODO()
+
+    return with(StringBuilder()) {
+        for (element in digit_storage) append(if (element < 10) "$element" else "${('a' + element - 10).toChar()}")
+        "$this"
```
Лучше вернуть из `with` просто `this`

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -184,7 +259,16 @@ fun convert(n: Int, base: Int): List<Int> = TODO()
+    if (n < 0 || base < 2 || base > 36) return ""
+
+    return with(StringBuilder()) {
+        for (element in digit_storage) append(if (element < 10) "$element" else "${('a' + element - 10).toChar()}")
```
Код, который делает из числа символ, лучше сделать отдельной функцией

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -204,7 +294,17 @@ fun decimal(digits: List<Int>, base: Int): Int = TODO()
+    var current_digit: Int
+
+    for (element in str) {
+        current_digit = (element - if ('0' <= element && element <= '9') '0' else 'a' - 10).toInt()
```
Код, который делает из символа число, лучше сделать отдельной функцией

ice-phoenix commented on src/lesson5/task1/Parse.kt:
```diff
@@ -60,7 +59,31 @@ fun main(args: Array<String>) {
+    }
+
+    return with(StringBuilder()) {
+        append(if (date_part_storage[0].length < 2) '0' + date_part_storage[0] else date_part_storage[0])
```
https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.text/pad-start.html

ice-phoenix commented on src/lesson5/task1/Parse.kt:
```diff
@@ -60,7 +59,31 @@ fun main(args: Array<String>) {
+        append('.')
+
+        val month_num = month_name_storage.indexOf(date_part_storage[1]) + 1
+        if (1 > month_num || month_num > 12) return ""
```
`month_num !in 1..12`

ice-phoenix commented on src/lesson5/task1/Parse.kt:
```diff
@@ -83,7 +106,20 @@ fun dateDigitToStr(digital: String): String = TODO()
+        if (!allowed_symbols.contains(target[index]))
+            return ""
+        else ++index
+    }
```
Вместо всего этого можно воспользоваться функцией `any{ ... } / all{ ... }`

ice-phoenix commented on src/lesson5/task1/Parse.kt:
```diff
@@ -95,7 +131,19 @@ fun flattenPhoneNumber(phone: String): String = TODO()
+    val lucky_jump_storage = mutableListOf <Int>()
+
+    try {
+        for (element in jump_storage) if (element != "-" && element != "%") lucky_jump_storage.add(element.toInt())
```
Все это можно выразить через `filter{ ... } + map { ... }`

ice-phoenix commented on src/lesson5/task1/Parse.kt:
```diff
@@ -95,7 +131,19 @@ fun flattenPhoneNumber(phone: String): String = TODO()
+    }
+
+    if (lucky_jump_storage.isEmpty()) return -1
+    return lucky_jump_storage.sortedDescending()[0]
```
Можно просто `lucky_jump_storage.max()`

ice-phoenix commented on src/lesson5/task1/Parse.kt:
```diff
@@ -107,7 +155,32 @@ fun bestLongJump(jumps: String): Int = TODO()
+fun bestHighJump(jumps: String): Int? {
+    // Проверка на кривые символы
+    val allowed_symbols = setOf('0', '1', '2', '3', '4', '5', '6', '7', '8', '9', ' ', '+', '-', '%')
+    if (jumps.any { !allowed_symbols.contains(it) }) return -1
```
Можно `it !in allowed_symbols`

DDRDmakar committed:
Code style changed everywhere

ice-phoenix commented on src/lesson1/task1/Simple.kt:
```diff
@@ -60,31 +61,31 @@ fun seconds(hours: Int, minutes: Int, seconds: Int): Int = TODO()
  * Например, расстояние между (3, 0) и (0, 4) равно 5
  */
-fun trackLength(x1: Double, y1: Double, x2: Double, y2: Double): Double = TODO()
+fun trackLength(x1: Double, y1: Double, x2: Double, y2: Double): Double = Math.sqrt(((x1 - x2) * (x1 - x2)) + ((y1 - y2) * (y1 - y2)))
```
Лишние скобки вокруг законченных выражений

ice-phoenix commented on src/lesson1/task1/Simple.kt:
```diff
@@ -102,12 +103,12 @@ fun travelMinutes(hoursDepart: Int, minutesDepart: Int, hoursArrive: Int, minute
  *Необходимо вывести число, полученное из заданного перестановкой цифр в обратном порядке (например, 874).
  */
-fun numberRevert(number: Int): Int = TODO()
+fun numberRevert(number: Int): Int = number%10 * 100 + number%100 - number%10 + (number - number%100) / 100
```
Форматирование кода

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -42,9 +52,21 @@ fun ageDescription(age: Int): String = TODO()
+
+    val half = (s1 + s2 + s3) / 2
+
+    // V T
```
Излишний комментарий

ice-phoenix commented on src/lesson2/task2/Logical.kt:
```diff
@@ -17,15 +18,23 @@ fun pointInsideCircle(x: Double, y: Double, x0: Double, y0: Double, r: Double) =
  */
-fun isNumberHappy(number: Int): Boolean = TODO()
+fun isNumberHappy(number: Int): Boolean {
+    val s1 = ((number - number % 1000) / 1000) + ((number % 1000 - number % 100) / 100)
```
Излишние скобки

ice-phoenix commented on src/lesson2/task2/Logical.kt:
```diff
@@ -46,4 +55,19 @@ fun circleInside(x1: Double, y1: Double, r1: Double,
+		if ( b < c ) min_2 = b
+		else min_2 = c
+	}
+	else {
```
`} else {`

ice-phoenix commented on src/lesson2/task2/Logical.kt:
```diff
@@ -46,4 +55,19 @@ fun circleInside(x1: Double, y1: Double, r1: Double,
  */
-fun brickPasses(a: Int, b: Int, c: Int, r: Int, s: Int): Boolean = TODO()
+fun brickPasses(a: Int, b: Int, c: Int, r: Int, s: Int): Boolean {
+	val min_1 : Int
```
`camelCase`

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,30 +58,63 @@ fun digitCountInNumber(n: Int, m: Int): Int =
  * минимальное число k, которое делится и на m и на n без остатка
  */
-fun lcm(m: Int, n: Int): Int = TODO()
+fun gcd(m: Int, n: Int): Int = if(n == 0) m else gcd(n, m % n)
```
Форматирование кода

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,30 +58,63 @@ fun digitCountInNumber(n: Int, m: Int): Int =
-fun lcm(m: Int, n: Int): Int = TODO()
+fun gcd(m: Int, n: Int): Int = if(n == 0) m else gcd(n, m % n)
+
+fun lcm(m: Int, n: Int): Int = m / gcd(m, n) *  n
```
Лучше `m * n / gcd(m, n)`

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -123,15 +176,45 @@ fun sin(x: Double, eps: Double): Double = TODO()
+    var part: Double
+    var i = 2
+    var total = 1.0
+    val xMin = x % (2 * Math.PI)
```
Почему в одной функции данная переменная названа `angle`, а в другой `xMin`?

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -104,22 +105,26 @@ fun buildSumExample(list: List<Int>) = list.joinToString(separator = " + ", post
+fun center(list: MutableList<Double>): MutableList<Double> {
+    if (list.isEmpty()) return list
+    val ariphmeticalАverage = mean(list)
+    return list.map { it - ariphmeticalАverage }.toMutableList()
```
Вы должны изменять список, переданный на вход

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -104,22 +105,26 @@ fun buildSumExample(list: List<Int>) = list.joinToString(separator = " + ", post
  */
-fun center(list: MutableList<Double>): MutableList<Double> = TODO()
+fun center(list: MutableList<Double>): MutableList<Double> {
+    if (list.isEmpty()) return list
```
Зачем эта проверка? Разве код ниже не будет работать для пустого списка?

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -157,15 +187,40 @@ fun accumulate(list: MutableList<Double>): MutableList<Double> = TODO()
-fun factorize(n: Int): List<Int> = TODO()
+fun numberIsPrime(n: Int): Boolean {
+    if (n == 0 || n == 1 || n == -1) return false
+    for (i in 2..Math.abs(n / 2)) if (n % i == 0) return false
```
Эту функцию можно написать проще, если вспомнить о других функциях, которые вы уже писали

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -157,15 +187,40 @@ fun accumulate(list: MutableList<Double>): MutableList<Double> = TODO()
+                currentNumber /= i
+                break
+            }
+        }
```
Может быть, весь этот цикл можно заменить на какую-то уже написанную вами функцию?

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -184,7 +251,18 @@ fun convert(n: Int, base: Int): List<Int> = TODO()
+    if (n < 0 || base < 2 || base > 36) return ""
+
+    with(StringBuilder()) {
+        for (element in digitStorage) append(numToSymbol(element))
```
Совсем круто было бы, если бы вы разобрались, как это написать через функцию `fold`

DDRDmakar committed:
Lesson 5 - mostExpensive

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -204,7 +288,17 @@ fun decimal(digits: List<Int>, base: Int): Int = TODO()
-fun decimalFromString(str: String, base: Int): Int = TODO()
+fun decimalFromString(str: String, base: Int): Int {
+    val digitStorage = mutableListOf <Int>()
+    var currentDigit: Int
```
Почему эта переменная объявлена здесь, а не в цикле?

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -204,7 +288,17 @@ fun decimal(digits: List<Int>, base: Int): Int = TODO()
+    var currentDigit: Int
+
+    for (element in str) {
+        currentDigit = (element - if ('0' <= element && element <= '9') '0' else 'a' - 10).toInt()
```
Лучше сделать отдельную функцию, которая переводит символ в число

ice-phoenix commented on src/lesson5/task1/Parse.kt:
```diff
@@ -60,7 +59,31 @@ fun main(args: Array<String>) {
+    }
+
+    return with(StringBuilder()) {
+        append(if (datePartStorage[0].length < 2) '0' + datePartStorage[0] else datePartStorage[0])
```
> Почитайте про функцию `String.padStart()`

Вы же ее и так используете ниже, почему вы не используете ее здесь???

ice-phoenix commented on src/lesson5/task1/Parse.kt:
```diff
@@ -83,7 +106,14 @@ fun dateDigitToStr(digital: String): String = TODO()
+    val target = phone
+    val allowedSymbols = setOf('0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '+', ' ', '-', '(', ')')
+
+    if(target.any { !allowedSymbols.contains(it) }) return ""
```
- Форматирование кода
- `a !in b`

ice-phoenix commented on src/lesson5/task1/Parse.kt:
```diff
@@ -95,7 +125,19 @@ fun flattenPhoneNumber(phone: String): String = TODO()
+    var luckyJumpStorage = listOf <Int>()
+
+    try {
+        luckyJumpStorage = (jumpStorage.filter { it != "-" && it != "%" }).map { it.toInt() }
```
Лишние скобки

ice-phoenix commented on src/lesson5/task1/Parse.kt:
```diff
@@ -107,7 +149,32 @@ fun bestLongJump(jumps: String): Int = TODO()
+    if (jumps.any { it !in allowedSymbols }) return -1
+
+    val jumpStorage = jumps.split(" ") // Разбиение на части
+    val attemptStorage = mutableListOf <Int>()
```
Форматирование кода

ice-phoenix commented on src/lesson5/task1/Parse.kt:
```diff
@@ -107,7 +149,32 @@ fun bestLongJump(jumps: String): Int = TODO()
+
+    // Формирование списка из удачных попыток
+    for (i in 0..jumpStorage.size - 1 step 2) {
+        if (jumpStorage[i + 1].any { '0' <= it && it <= '9' })
```
`it in ...`

ice-phoenix commented on src/lesson5/task1/Parse.kt:
```diff
@@ -118,7 +185,32 @@ fun bestHighJump(jumps: String): Int = TODO()
+        throw IllegalArgumentException("null")
+    }
+
+    return target
```
[Здесь и везде] Внутри `try` можно писать, в том числе, и `return`

ice-phoenix commented on src/lesson5/task1/Parse.kt:
```diff
@@ -118,7 +185,32 @@ fun bestHighJump(jumps: String): Int = TODO()
+                target += expressionStorage[i+1].toInt()
+            }
+            else if(expressionStorage[i] == "-") {
+                target -= expressionStorage[i+1].toInt()
```
Форматирование кода

ice-phoenix commented on src/lesson5/task1/Parse.kt:
```diff
@@ -142,7 +234,29 @@ fun firstDuplicateIndex(str: String): Int = TODO()
+            separator = element.lastIndexOf(" ")
+            currentPrice = element.drop(separator + 1).toDouble()
+            if (currentPrice >= maxPrice) {
+                name = element.dropLast(element.length - separator)
```
Можно `take(separator)`

DDRDmakar committed:
style enhanced (all lessons)

DDRDmakar committed:
'center' fixed

DDRDmakar committed:
'mean' fixed

DDRDmakar committed:
beginning of geometry

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,30 +58,63 @@ fun digitCountInNumber(n: Int, m: Int): Int =
+fun fib(n: Int): Int {
+    var f = 1
+    var fPrev = 1
+    var fTemp: Int
```
Почему не проинициализировать эту переменную каким-либо значением?

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,30 +58,63 @@ fun digitCountInNumber(n: Int, m: Int): Int =
+    if (n == 1 || n == 2) return 1
+
+    for (i in 3..n) {
+        fTemp = f
```
Почему не объявить эту переменную прямо тут?

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -114,7 +148,26 @@ fun squareBetweenExists(m: Int, n: Int): Boolean = TODO()
+        part = Math.pow(angle, i.toDouble()) / factorial(i)
+        i += 2
+        total -= part
+        if(Math.abs(eps) > Math.abs(part)) break
```
Форматирование кода

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -104,22 +107,27 @@ fun buildSumExample(list: List<Int>) = list.joinToString(separator = " + ", post
-fun center(list: MutableList<Double>): MutableList<Double> = TODO()
+fun center(list: MutableList<Double>): MutableList<Double> {
+    val d = mean(list)
+    for((index, element) in list.withIndex())
```
Форматирование кода

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -157,15 +192,34 @@ fun accumulate(list: MutableList<Double>): MutableList<Double> = TODO()
+    var currentNumber = n
+    val primeDivisorsStorage = mutableListOf <Int>()
+
+    while (!numberIsPrime(currentNumber)) {
```
Нельзя ли обойтись без функции `numberIsPrime`?

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -174,7 +228,19 @@ fun factorizeToString(n: Int): String = TODO()
  */
-fun convert(n: Int, base: Int): List<Int> = TODO()
+fun convert(n: Int, base: Int): List<Int> {
+    val digitStorage = mutableListOf <Int>()
```
Форматирование кода

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -174,7 +228,19 @@ fun factorizeToString(n: Int): String = TODO()
+    var currentNumber = n
+
+    if (n == 0) return listOf(0)
+    if (n < 0 || base < 2) return listOf()
```
Почему вы проверяете вход на ошибки после того, как обработали один отдельный, но корректный случай?

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -204,7 +286,12 @@ fun decimal(digits: List<Int>, base: Int): Int = TODO()
+
+fun decimalFromString(str: String, base: Int): Int {
+    val digitStorage = str.map { symbolToNum(it) }
+    return decimal(digitStorage, base)
```
:+1:

ice-phoenix commented on src/lesson5/task1/Parse.kt:
```diff
@@ -60,7 +59,35 @@ fun main(args: Array<String>) {
+
+        val monthNum = monthNameStorage.indexOf(datePartStorage[1]) + 1
+        if (monthNum !in 1..12)
+            return ""
```
Эту проверку лучше делать до того, как вы начинаете собирать строку в `StringBuilder`

ice-phoenix commented on src/lesson5/task1/Parse.kt:
```diff
@@ -95,7 +126,19 @@ fun flattenPhoneNumber(phone: String): String = TODO()
+    }
+
+    if (luckyJumpStorage.isEmpty()) return -1
+    return luckyJumpStorage.max()
```
Можно все это выразить проще через `?:`

ice-phoenix commented on src/lesson5/task1/Parse.kt:
```diff
@@ -107,7 +150,31 @@ fun bestLongJump(jumps: String): Int = TODO()
+    if (jumps.any { it !in "0123456789 +-%" }) return -1
+
+    val jumpStorage = jumps.split(" ") // Разбиение на части
+    val attemptStorage = mutableListOf <Int>()
```
Форматирование кода

ice-phoenix commented on src/lesson5/task1/Parse.kt:
```diff
@@ -107,7 +150,31 @@ fun bestLongJump(jumps: String): Int = TODO()
+    if (attemptStorage.isEmpty()) return -1
+
+    // Выбор максимального значения
+    return attemptStorage.max()
```
См. выше

ice-phoenix commented on src/lesson5/task1/Parse.kt:
```diff
@@ -118,7 +185,28 @@ fun bestHighJump(jumps: String): Int = TODO()
-fun plusMinus(expression: String): Int = TODO()
+fun plusMinus(expression: String): Int {
+    val expressionStorage = expression.split(" ").filter { !it.isEmpty() }
+    var target: Int
```
Почему эта переменная объявлена здесь?

ice-phoenix commented on src/lesson5/task1/Parse.kt:
```diff
@@ -142,7 +230,28 @@ fun firstDuplicateIndex(str: String): Int = TODO()
-fun mostExpensive(description: String): String = TODO()
+fun mostExpensive(description: String): String {
+    val goods = description.split("; ")
+    var separator: Int
```
Почему эта переменная объявлена здесь?

ice-phoenix commented on src/lesson5/task1/Parse.kt:
```diff
@@ -142,7 +230,28 @@ fun firstDuplicateIndex(str: String): Int = TODO()
+    val goods = description.split("; ")
+    var separator: Int
+    var name = ""
+    var currentPrice: Double
```
Почему эта переменная объявлена здесь?

ice-phoenix commented on src/lesson6/task1/Geometry.kt:
```diff
@@ -55,28 +56,41 @@ data class Circle(val center: Point, val radius: Double) {
  */
-data class Segment(val begin: Point, val end: Point)
+data class Segment(val begin: Point, val end: Point) {
+    fun length(): Double = begin.distance(end)
```
Если вам необходимо расширить сторонний класс, в Котлине для этого используются `extension methods`, а не `я забрался в код и поменял все как мне надо`

ice-phoenix replied:
Все еще актуально

ice-phoenix commented:
recheck all

DDRDmakar committed:
lessons 1 - 5 fixes

ice-phoenix commented on src/lesson1/task1/Simple.kt:
```diff
@@ -93,7 +94,7 @@ fun thirdDigit(number: Int): Int = TODO()
  * Определите время поезда в пути в минутах (в данном случае 216).
  */
-fun travelMinutes(hoursDepart: Int, minutesDepart: Int, hoursArrive: Int, minutesArrive: Int): Int = TODO()
+fun travelMinutes(hoursDepart: Int, minutesDepart: Int, hoursArrive: Int, minutesArrive: Int): Int = hoursArrive * 60 + minutesArrive - hoursDepart * 60 - minutesDepart
```
Длинные строки

ice-phoenix replied:
Все еще актуально

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -42,9 +52,19 @@ fun ageDescription(age: Int): String = TODO()
+
+    return when {
+        (half <= s1) -> half / v1
+        (half > s1 && half <= s1 + s2) -> t1 + ((half - s1) / v2)
```
Первую часть можно не проверять, вы ее уже проверили строчкой выше

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -89,4 +120,20 @@ fun triangleKind(a: Double, b: Double, c: Double): Int = TODO()
+    // ========[c]######[a]|||||||[b]######[d]=========> b - a
+    // ========[c]######[d]=======[a]######[b]=========> -
+
+    return when {
```
Функцию с телом в виде одного `return expr` можно записать как `fun foo(...) = expr`

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -89,4 +120,20 @@ fun triangleKind(a: Double, b: Double, c: Double): Int = TODO()
+        (a <= c && c <= b && b <= d) -> b - c
+        (a <= c && c <= d && d <= b) -> d - c
+        (c <= a && a <= d && d <= b) -> d - a
+        (c <= a && a <= b && b <= d) -> b - a
```
Данный набор условий можно записать намного короче, если подумать, как связаны уменьшаемое и вычитаемое с границами отрезков и функциями `min / max`

ice-phoenix replied:
Все еще актуально

ice-phoenix commented on src/lesson2/task2/Logical.kt:
```diff
@@ -46,4 +55,18 @@ fun circleInside(x1: Double, y1: Double, r1: Double,
+		min1 = b
+		if ( a < c ) min2 = a
+		else min2 = c
+	}
```
Форматирование кода

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,30 +58,62 @@ fun digitCountInNumber(n: Int, m: Int): Int =
+    do {
+        ++digitsAmount
+        currentNumber /= 10
+
```
Зачем эта пустая строка?

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -57,30 +58,62 @@ fun digitCountInNumber(n: Int, m: Int): Int =
+    var f = 1
+    var fPrev = 1
+
+
```
Многовато пустых строк

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -156,8 +246,20 @@ fun hasDifferentDigits(n: Int): Boolean = TODO()
+
+    for (i in 1..n) {
+        lastSequenceElement = "${i * i}"
+        if (sequenceLen + lastSequenceElement.length >= n) return (lastSequenceElement[n - sequenceLen - 1] - '0').toInt()
```
Длинная строка

ice-phoenix commented on src/lesson3/task1/Loop.kt:
```diff
@@ -156,8 +246,20 @@ fun hasDifferentDigits(n: Int): Boolean = TODO()
+        sequenceLen += lastSequenceElement.length
+    }
+
+    return (lastSequenceElement [n - 1] - '0').toInt()
```
Форматирование кода

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -157,15 +192,33 @@ fun accumulate(list: MutableList<Double>): MutableList<Double> = TODO()
  * Множители в списке должны располагаться по возрастанию.
  */
-fun factorize(n: Int): List<Int> = TODO()
+fun numberIsPrime(n: Int): Boolean {
```
Зачем вам эта функция?

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -174,7 +227,19 @@ fun factorizeToString(n: Int): String = TODO()
+    if (n < 0 || base < 2) return listOf()
+    if (n == 0) return listOf(0)
+
+    val digitStorage = mutableListOf <Int>()
```
Форматирование кода

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -214,7 +298,20 @@ fun decimalFromString(str: String, base: Int): Int = TODO()
  */
-fun roman(n: Int): String = TODO()
+fun roman(n: Int): String {
+    val digit_1 = listOf("", "I", "II", "III", "IV", "V", "VI", "VII", "VIII", "IX") // 1 2 3 4 5 6 7 8 9
```
`camelCase`

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -214,7 +298,20 @@ fun decimalFromString(str: String, base: Int): Int = TODO()
+        append(digit_10[(n % 100) / 10])
+        append(digit_1[n % 10])
+
+        return this.toString()
```
:+1:

ice-phoenix commented on src/lesson5/task1/Parse.kt:
```diff
@@ -95,7 +121,18 @@ fun flattenPhoneNumber(phone: String): String = TODO()
+        return -1
+    }
+
+    return if (luckyJumpStorage.isEmpty()) -1 else luckyJumpStorage?.max() ?: -1
```
Зачем первая проверка?

ice-phoenix commented on src/lesson5/task1/Parse.kt:
```diff
@@ -95,7 +121,18 @@ fun flattenPhoneNumber(phone: String): String = TODO()
+    val luckyJumpStorage: List<Int>
+
+    try {
+        luckyJumpStorage = jumpStorage.filter { it !in "-%" }.map { it.toInt() }
```
Почему вы разделили определение и объявление переменной?

ice-phoenix commented on src/lesson5/task1/Parse.kt:
```diff
@@ -107,7 +144,32 @@ fun bestLongJump(jumps: String): Int = TODO()
+        return -1
+
+    val jumpStorage = jumps.split(" ") // Разбиение на части
+    val attemptStorage = mutableListOf <Int>()
```
Форматирование кода

ice-phoenix commented on src/lesson5/task1/Parse.kt:
```diff
@@ -107,7 +144,32 @@ fun bestLongJump(jumps: String): Int = TODO()
+    }
+
+    // Выбор максимального значения
+    return if (attemptStorage.isEmpty()) -1 else attemptStorage.max()
```
Лучше через элвис-оператор

ice-phoenix commented on src/lesson5/task1/Parse.kt:
```diff
@@ -118,7 +180,27 @@ fun bestHighJump(jumps: String): Int = TODO()
+                target += expressionStorage[i + 1].toInt()
+            } else if (expressionStorage[i] == "-") {
+                target -= expressionStorage[i + 1].toInt()
+            } else throw IllegalArgumentException("null")
```
Зачем `"null"`?

DDRDmakar replied:
Конструктор требует строку. Что нужно ему передавать? Я не знаю, поэтому передаю "null"

ice-phoenix commented on src/lesson6/task1/Geometry.kt:
```diff
@@ -55,28 +56,41 @@ data class Circle(val center: Point, val radius: Double) {
+            if (e1.distance(e2) > longestSegment.length()) longestSegment = Segment(e1, e2)
+        }
+    }
+    return longestSegment
```
Как ваш код проверяет, что в множестве менее двух точек?

DDRDmakar replied:
Я исправил не всё, так и написал в названрии коммита.

ice-phoenix replied:
Все еще актуально

ice-phoenix commented:
recheck all

DDRDmakar committed:
Fixes + Chess started

DDRDmakar committed:
rookMoveNumber and rookTrajectory fixed

DDRDmakar committed:
lesson 7 task 1

DDRDmakar committed:
lesson 7 task 1

DDRDmakar committed:
lesson 7 task 2

DDRDmakar committed:
lesson 7 task 1

ice-phoenix commented:
recheck all

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -157,15 +192,27 @@ fun accumulate(list: MutableList<Double>): MutableList<Double> = TODO()
+        currentNumber /= minDivisorValue
+    }
+
+    return primeDivisorsStorage.sorted()
```
Зачем сортировать?

ice-phoenix commented on src/lesson5/task1/Parse.kt:
```diff
@@ -95,7 +121,15 @@ fun flattenPhoneNumber(phone: String): String = TODO()
+    try {
+        val jumpStorage = jumps.split(" ")
+        val luckyJumpStorage = jumpStorage.filter { it !in "-%" }.map { it.toInt() }
+        return luckyJumpStorage?.max() ?: -1
```
Зачем тут safe call в `list?.max()`?

DDRDmakar replied:
Потому что max() возвращает Int?

ice-phoenix replied:
Это прекрасно, но зачем делать safe call на том, что вполне себе не является nullable типом?

DDRDmakar replied:
Иначе отказывается компилироваться

ice-phoenix replied:
Не верю (с) Обсудите с преподавателем на занятиях, чтобы он объяснил, что и как

ice-phoenix commented on src/lesson6/task1/Geometry.kt:
```diff
@@ -98,28 +115,51 @@ data class Line(val point: Point, val angle: Double) {
-fun bisectorByPoints(a: Point, b: Point): Line = TODO()
+fun bisectorByPoints(a: Point, b: Point): Line {
+    val m = Segment(a, b).mid()
+    val ang = Math.atan((b.y - a.y) / (b.x - a.x)) + Math.PI / 2
```
Переменным лучше давать осмысленные имена, соответствующие их назначению в коде

ice-phoenix commented on src/lesson6/task2/Chess.kt:
```diff
@@ -23,7 +23,9 @@ data class Square(val column: Int, val row: Int) {
      */
-    fun notation(): String = TODO()
+    fun notation(): String {
+        return if(inside()) ('a' + column - 1).toString() + row.toString() else ""
```
Функцию с телом в виде одного `return expr` можно записать как `fun foo(...) = expr`

ice-phoenix replied:
Форматирование кода

ice-phoenix commented on src/lesson6/task2/Chess.kt:
```diff
@@ -33,7 +35,10 @@ data class Square(val column: Int, val row: Int) {
  */
-fun square(notation: String): Square = TODO()
+fun square(notation: String): Square {
+    if(!notation.matches(Regex("""[a-h][1-8]"""))) throw IllegalArgumentException()
```
Форматирование кода

ice-phoenix commented on src/lesson6/task2/Chess.kt:
```diff
@@ -58,7 +63,16 @@ fun square(notation: String): Square = TODO()
-fun rookMoveNumber(start: Square, end: Square): Int = TODO()
+fun Square.onLineWith(other: Square): Boolean = column == other.column || row == other.row
+
+fun Square.similarWith(other: Square): Boolean = column == other.column && row == other.row
```
Как мне кажется, это просто операция равенства для клеток

ice-phoenix commented on src/lesson6/task2/Chess.kt:
```diff
@@ -58,7 +63,16 @@ fun square(notation: String): Square = TODO()
  * Ладья может пройти через клетку (3, 3) или через клетку (6, 1) к клетке (6, 3).
  */
-fun rookMoveNumber(start: Square, end: Square): Int = TODO()
+fun Square.onLineWith(other: Square): Boolean = column == other.column || row == other.row
```
:+1:

ice-phoenix commented on src/lesson7/task1/Matrix.kt:
```diff
@@ -38,32 +55,74 @@ interface Matrix<E> {
-    override val height: Int = TODO()
+class MatrixImpl<E>(override val height: Int, override val width: Int, e: E) : Matrix<E> {
+
+    private fun createRow(length: Int, e: E) = Array<Any?>(length) { j -> e }
```
Почему `Any?`?

DDRDmakar replied:
Потому что компилятор не принимает массив шаблонов. 
" Cannot use 'T' as refied type parameter. Use a class instead."

ice-phoenix commented on src/lesson7/task1/Matrix.kt:
```diff
@@ -38,32 +55,74 @@ interface Matrix<E> {
 
     override fun set(row: Int, column: Int, value: E) {
-        TODO()
+        if (row !in 0..height - 1 || column !in 0..width - 1) return
```
Почему здесь вы проверяете границы, а в `get` --- нет?

ice-phoenix commented on src/lesson7/task1/Matrix.kt:
```diff
@@ -38,32 +55,74 @@ interface Matrix<E> {
+            stringStorage.append("    ")
+            for (j in 0..width - 1) {
+                stringStorage.append("[${storage[i][j]}]")
+                if (j != width - 1) stringStorage.append(",")
```
Чтобы избежать дорогой проверки внутри цикла, обычно разделитель добавляют всегда, а снаружи цикла откусывают лишний

ice-phoenix commented on src/lesson7/task2/Matrices.kt:
```diff
@@ -59,7 +60,41 @@ operator fun Matrix<Int>.plus(other: Matrix<Int>): Matrix<Int> {
+    var i = 0
+    var j = 0
+    var spiralNumber = 1
+    while (matrixMin(field) == 0) {
```
Может быть, условие завершения можно сделать попроще, чем каждый раз обходить ВСЮ матрицу и искать в ней нули?

ice-phoenix commented on src/lesson7/task2/Matrices.kt:
```diff
@@ -118,7 +153,31 @@ fun <E> rotate(matrix: Matrix<E>): Matrix<E> = TODO()
+        }
+
+        println(column)
+        println(row)
```
Это еще что такое?

ice-phoenix commented on src/lesson7/task2/Matrices.kt:
```diff
@@ -118,7 +153,31 @@ fun <E> rotate(matrix: Matrix<E>): Matrix<E> = TODO()
+        // Пихаю в Set по одной строке и одному столбцу
+        val row = mutableSetOf<Int>()
+        val column = mutableSetOf<Int>()
+        for(j in 0..D - 1) {
```
Форматирование кода

ice-phoenix commented on src/lesson7/task2/Matrices.kt:
```diff
@@ -118,7 +153,31 @@ fun <E> rotate(matrix: Matrix<E>): Matrix<E> = TODO()
+            column.any { it !in 1..D } ||
+            row.any { it !in 1..D } ||
+            column.size != D ||
+            row.size != D
```
Это все можно проверить просто через равенство ваших множеств какому-то эталонному

mglukhikh commented:
recheck all

DDRDmakar committed:
Half fixed

ice-phoenix commented:
recheck all

DDRDmakar committed:
L7

DDRDmakar committed:
L7

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -89,4 +120,15 @@ fun triangleKind(a: Double, b: Double, c: Double): Int = TODO()
-fun segmentLength(a: Int, b: Int, c: Int, d: Int): Int = TODO()
\ No newline at end of file
+fun segmentLength(a: Int, b: Int, c: Int, d: Int): Int {
+    val min_coordinate = listOf<Int>(a, b, c, d).min()
```
`camelCase`

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -89,4 +120,15 @@ fun triangleKind(a: Double, b: Double, c: Double): Int = TODO()
\ No newline at end of file
+fun segmentLength(a: Int, b: Int, c: Int, d: Int): Int {
+    val min_coordinate = listOf<Int>(a, b, c, d).min()
+    val max_coordinate = listOf<Int>(a, b, c, d).max()
```
Это совсем не упрощает набор условий

ice-phoenix commented on src/lesson2/task2/Logical.kt:
```diff
@@ -46,4 +55,18 @@ fun circleInside(x1: Double, y1: Double, r1: Double,
+		else min2 = c
+	}
+	
+	return (min1 <= r && min2 <= s) || (min1 <= s && min2 <= r)
```
Форматирование кода

ice-phoenix commented on src/lesson7/task1/Matrix.kt:
```diff
@@ -38,32 +55,81 @@ interface Matrix<E> {
  */
-fun <E> createMatrix(height: Int, width: Int, e: E): Matrix<E> = TODO()
+fun <E> createMatrix(height: Int, width: Int, e: E): Matrix<E> {
+    if(height <= 0 || width <= 0) throw IllegalArgumentException()
```
Форматирование кода

ice-phoenix commented on src/lesson7/task1/Matrix.kt:
```diff
@@ -38,32 +55,81 @@ interface Matrix<E> {
+class MatrixImpl<E>(override val height: Int, override val width: Int, e: E) : Matrix<E> {
 
-    override val width: Int = TODO()
+    private fun createRow(length: Int, e: E) = Array<Any?>(length) { j -> e }
```
Тип у массива можно указать точнее, чем `Any?`

mglukhikh replied:
Просто для информации (возможно, в вопросе речь не об этом): попытка создать `Array<E>` приведёт к ошибке компиляции, так как массивы в JVM имеют reified type parameters.

ice-phoenix commented on src/lesson7/task1/Matrix.kt:
```diff
@@ -38,32 +55,81 @@ interface Matrix<E> {
 
     override fun set(row: Int, column: Int, value: E) {
-        TODO()
+        if (row !in 0..height - 1 || column !in 0..width - 1) return
```
Почему при чтении элемента выход за границу --- это исключение, а при записи --- нет?

ice-phoenix commented on src/lesson7/task1/Matrix.kt:
```diff
@@ -38,32 +55,81 @@ interface Matrix<E> {
+    override fun hashCode(): Int {
+        var res = 9
+        for (elem in storage) {
+            res = res * 23 + elem.hashCode()
```
Почему хэш код не учитывает размеры матрицы?

ice-phoenix commented on src/lesson7/task2/Matrices.kt:
```diff
@@ -59,7 +60,35 @@ operator fun Matrix<Int>.plus(other: Matrix<Int>): Matrix<Int> {
+        ++i
+        ++j
+
+        cellNumber += 4
```
Зачем вам эта переменная?

ice-phoenix commented on src/lesson7/task2/Matrices.kt:
```diff
@@ -118,7 +147,26 @@ fun <E> rotate(matrix: Matrix<E>): Matrix<E> = TODO()
+fun isLatinSquare(matrix: Matrix<Int>): Boolean {
+    if (matrix.height != matrix.width) return false
+    val D = matrix.height
+    val etalon = mutableSetOf<Int>()
```
Называть переменные транслитом --- не очень хорошо

DDRDmakar committed:
L 1-7 [2]

DDRDmakar committed:
L 7 [2]

DDRDmakar committed:
L 7 [2] fixes

ice-phoenix commented:
recheck all

ice-phoenix commented:
recheck all

ice-phoenix commented on src/lesson2/task1/IfElse.kt:
```diff
@@ -89,4 +120,8 @@ fun triangleKind(a: Double, b: Double, c: Double): Int = TODO()
-fun segmentLength(a: Int, b: Int, c: Int, d: Int): Int = TODO()
\ No newline at end of file
+fun segmentLength(a: Int, b: Int, c: Int, d: Int): Int {
+    var sumLen = setOf(b, d)?.min() ?: 0
```
Почему бы не воспользоваться просто функциями `Math.min / Math.max`?

ice-phoenix commented on src/lesson4/task1/List.kt:
```diff
@@ -157,15 +192,27 @@ fun accumulate(list: MutableList<Double>): MutableList<Double> = TODO()
-fun factorize(n: Int): List<Int> = TODO()
+
+fun factorize(n: Int): List<Int> {
+    val primeDivisorsStorage = mutableListOf <Int>()
```
Форматирование кода

ice-phoenix commented on src/lesson7/task1/Matrix.kt:
```diff
@@ -38,32 +55,82 @@ interface Matrix<E> {
+    }
+
+    override fun get(cell: Cell): E  {
+        if (cell.row !in 0..height - 1 || cell.column !in 0..width - 1) throw IllegalArgumentException()
```
Почему просто не вызвать метод выше?

ice-phoenix commented on src/lesson7/task1/Matrix.kt:
```diff
@@ -38,32 +55,82 @@ interface Matrix<E> {
 
     override fun set(cell: Cell, value: E) {
-        TODO()
+        if (cell.row !in 0..height - 1 || cell.column !in 0..width - 1) throw IllegalArgumentException()
```
Почему просто не вызвать метод выше?

ice-phoenix commented on src/lesson7/task1/Matrix.kt:
```diff
@@ -38,32 +55,82 @@ interface Matrix<E> {
+
+        for (i in 0..height - 1) {
+            for (j in 0..width - 1) {
+               if (storage[i][j] != other.storage[i][j]) return false
```
Весь этот цикл можно заменить на `==`

ice-phoenix commented on src/lesson7/task2/Matrices.kt:
```diff
@@ -137,7 +210,24 @@ fun isLatinSquare(matrix: Matrix<Int>): Boolean = TODO()
+            if (i < matrix.height - 1 && j < matrix.width - 1) field[i, j] += matrix[i+1, j+1]
+            if (j < matrix.width - 1) field[i, j] += matrix[i, j+1]
+            if (i > 0 && j < matrix.width - 1) field[i, j] += matrix[i-1, j+1]
+            if (i > 0) field[i, j] += matrix[i-1, j]
```
Все это можно записать много проще, если расщепить один цикл на несколько --- по циклу на границу матрицы, плюс еще один по внутренним элементам

ice-phoenix commented:
С учетом статистики решений у вас все очень "Хорошо". Поздравляем!

DDRDmakar committed:
L 7 [3] fixes

Commit:
commit message

Commit:
commit message 2
